geth/node/manager.go

```diff
@@ -70,13 +76,91 @@ func (m *NodeManager) startNode(config *params.NodeConfig) (<-chan struct{}, err
 		return nil, ErrNodeExists
 	}
 
-	ethNode, err := MakeNode(config)
+	m.nodeSyncStarted = make(chan struct{}, 0)
+	m.nodeSyncError = make(chan struct{}, 0)
+	m.nodeSyncCompleted = make(chan struct{}, 0)
+
+	ethNode, err := MakeNode(config, func(leth *les.LightEthereum, stack *node.Node, config *params.NodeConfig) {
+		downlder := leth.Downloader()
+
+		go func() {
+			syncStateWait := time.NewTimer(params.DelayCycleForSyncStart)
+			syncStateWaitMax := time.NewTimer(params.DelayCycleForSyncStartMaxWait)
+
+			defer syncStateWait.Stop()
+			defer syncStateWaitMax.Stop()
+
+			// Initially we want to listen for when node have started synchronizing it's data.
+		nloop:
+			for {
+				select {
+				case <-m.nodeSyncError:
+					// Listen for error to ensure we are not in a case of concurrent runs of this,
+					// as I noticed an issue of double closure of channels, but no way of knowing why.
+					// This provides us a benefit that the go-routine will instantly be shutdown if the channel
+					// was already closed.
+					return
+				case <-syncStateWaitMax.C:
+					close(m.nodeSyncError)
+					return
+				case <-syncStateWait.C:
+					if downlder.Synchronising() {
+						close(m.nodeSyncStarted)
+						break nloop
+					}
+					syncStateWait.Reset(params.DelayCycleForSyncStart)
+				}
+			}
+
+			syncStateWait.Reset(params.DelayCycleForSyncCompleted)
+			syncStateWaitMax.Reset(params.DelayCycleForSyncCompletedMaxWait)
+
+			// Once we have validating, that things have properly started then, we then track
+			// progress for when a finished signal is called.
+		cloop:
+			for {
+				select {
+				case <-m.nodeSyncError:
+					// Listen for error to ensure we are not in a case of concurrent runs of this,
+					// as I noticed an issue of double closure of channels, but no way of knowing why.
+					// This provides us a benefit that the go-routine will instantly be shutdown if the channel
+					// was already closed.
+					return
+				case <-syncStateWaitMax.C:
+					close(m.nodeSyncError)
+					return
+				case <-syncStateWait.C:
+					progress := downlder.Progress()
+					if progress.CurrentBlock >= progress.HighestBlock {
+						close(m.nodeSyncCompleted)
+						break cloop
+					}
+					syncStateWait.Reset(params.DelayCycleForSyncCompleted)
+				}
+			}
+		}()
+	})
+
 	if err != nil {
+		close(m.nodeSyncError)
 		return nil, err
 	}
 
 	m.nodeStarted = make(chan struct{}, 1)
 
+	// If we are expected to use external RPC upstream connection, then
+	// immediately close synchronization notification channel.
+	if config.UpstreamConfig.Enabled {
+		close(m.nodeSyncStarted)
```

Comment:
Not sure it's a good idea, because it's wrong. It has never started syncing and we tell here that it not only started but also finished at the same time. As I proposed these `SyncDone()` and `SyncErr()` methods, we could return a specific error like `errors.New("sync not available because using upstream node)`.

Reply:
Hmmm, interesting. But issue here is, we must still expose an API method like `WaitOnSync()` else we wont guarantee sync or must ensure the places needed to poll for when `SyncDone() bool` returns true.

The idea is provide methods that can help us guarantee that sync was finished hence the use of signal channels. 

Else we should then have `SyncDone` block, and have them if any error occurs with `SyncErr` ?